reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java

```diff
@@ -0,0 +1,404 @@
+
+import static reactor.core.Fuseable.ConditionalSubscriber;
+
+/**
+ * @author Oleh Dokuka
+ */
+public class FluxLimitRate<T> extends InternalFluxOperator<T, T> {
+
+    final int highTide;
+
+    final int lowTide;
+
+    FluxLimitRate(Flux<? extends T> source, int highTide, int lowTide) {
+        super(source);
+        if (highTide <= 0) {
+            throw new IllegalArgumentException("highTide > 0 required but it was " + highTide);
+        }
+        this.highTide = highTide;
+        this.lowTide = lowTide;
+    }
+
+    @Override
+    public CoreSubscriber<? super T> subscribeOrReturn(CoreSubscriber<? super T> actual) {
+        if (actual instanceof ConditionalSubscriber) {
+            @SuppressWarnings("unchecked") // javac, give reason to suppress because inference anomalies
+            final ConditionalSubscriber<? super T> cs =
+                (ConditionalSubscriber<? super T>) actual;
+          return new LimitRateConditionalSubscriber<>(cs, highTide, lowTide);
+        }
+        return new LimitRateSubscriber<>(actual, highTide, lowTide);
+    }
+
+    private final static class LimitRateSubscriber<T> implements InnerOperator<T, T> {
+
+        private final int highTide;
+        private final int limit;
+
+        volatile long requested; // need sync
+        static final AtomicLongFieldUpdater<LimitRateSubscriber> REQUESTED =
+            AtomicLongFieldUpdater.newUpdater(LimitRateSubscriber.class, "requested");
+
+        volatile int wip; // need sync
+        static final AtomicIntegerFieldUpdater<LimitRateSubscriber> WIP =
+            AtomicIntegerFieldUpdater.newUpdater(LimitRateSubscriber.class, "wip");
+
+        volatile int produced = 0;
+        static final AtomicIntegerFieldUpdater<LimitRateSubscriber> PRODUCED =
+            AtomicIntegerFieldUpdater.newUpdater(LimitRateSubscriber.class, "produced");
+
+        private int pending;
+        // need sync since should be checked/zerroed in onNext
+        // and increased in request
+        private int counter; // no need to sync since increased zerroed only in
+        // the request method
+
+        final CoreSubscriber<? super T> actual;
+
+        Subscription s;
+
+        private LimitRateSubscriber(CoreSubscriber<? super T> actual,
+            int highTide,
+            int lowTide) {
+            this.actual = actual;
+            this.highTide = highTide;
+            this.limit = Operators.unboundedOrLimit(highTide, lowTide);
+        }
+
+        @Override
+        public void request(long n) {
+            if (Operators.validate(n)) {
+                Operators.addCap(REQUESTED, this, n);
+
+                tryRequest();
+            }
+        }
+
+        void tryRequest() {
+            if (WIP.getAndIncrement(this) != 0) {
+                return;
+            }
+
+
+            final Subscription s = this.s;
+            final int highTide = this.highTide;
+
+            int missed = 1;
+            int pending = this.pending;
```

Comment:
I think the `pending`, `produced` and `deliver` variable could benefit from either a renaming or a comment about what they represent, since several operations are performed on them that are not trivial to follow.